Integrate with a CSS parser for selector parsing

aweary opened the pull request:
Resolves #694, #693, #547, #1082, #1019, #1078, #1086, #283

This PR integrates with [rst-selector-parser](https://github.com/aweary/rst-selector-parser) a fork of [scalpel](https://github.com/gajus/scalpel/) which is a CSS parser implemented with [nearley](https://nearley.js.org/). This fork implements a custom grammar requiring quoted string attribute values, allowing unquoted numbers, booleans, and other primitives, and supports selectors in any order.

# Major Changes

* `buildPredicate` is exported by the `selectors` module, the API remains the same

* `ComplexSelector` is no longer needed. Instead the `selectors` module exports a `reduceTreeBySelector` which manages traversal + matching with simple selectors and combinators. Both the shallow and mount wrappers use the same `buildPredicate` and `findWhereUnwrapped` API so there's no need to have a class for selector parsing.


# Pending Changes

These are changes that I still need to make before this can be merged.

- [x] Remove `ComplexSelector` and all utility methods that were only used by `ComplexSelector`
- [x] Throw if a complex selector is passed to `buildPredicate`
- [x] Handle `PSEUDO_CLASS` and `PSEUDO_ELEMENT` selectors
- [x] Add tests for valid selectors that were previously unparsable

aweary commented on packages/enzyme-test-suite/test/ReactWrapper-spec.jsx:
```diff
@@ -433,22 +435,15 @@ describeWithDOM('mount', () => {
     });
 
-    // React 15.2 warns when setting a non valid prop to an DOM element
-    describeIf(REACT013 || REACT014, 'unauthorized dom props', () => {
```
This test was kind of a false positive. You could still query unauthorized DOM props by using the object syntax, and AFAICT it was only failing to find them because of how it was parsing the selector.

React <0.14 didn't just ignore syntactically invalid attributes, it ignored attributes that it didn't have whitelisted. To implement this behavior as it should be implemented we'd need to integrate that same whitelist.

ljharb replied:
Can we add that whitelist in the react 0.14 and 0.13 adapters?

aweary replied:
Sure, OK to do in a follow up?

ljharb replied:
Yes, totes ok to do in a followup

ljharb commented on packages/enzyme/src/Utils.js:
```diff
@@ -308,14 +308,6 @@ export function coercePropValue(propName, propValue) {
 
   const trimmedValue = propValue.trim();
 
-  // if propValue includes quotes, it should be
```
Why? The fact that selectors can omit quotes is imo very problematic; enforcing strings ensures that selectors used will be unambiguous.

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -155,13 +153,6 @@ describe('RSTTraversal', () => {
       expect(nodeHasProperty(<div foo={0} />, 'foo', '-Infinity')).to.equal(false);
     });
-
-    it('should throw when un unquoted string is passed in', () => {
```
tagging this line too - i think the purpose of requiring strings here was so people would distinguish between numbers and strings, for example.

Regardless of what mistakes React 16 has made, older Reacts do, and non-Reacts might, still want to distinguish these on host components - and they'll always need distinguishing on custom components.

aweary replied:
According to the spec, unquoted values must be valid identifiers or strings. The selectors `[foo=bar]` and `[foo="bar"]` are thus parsed identically. A spec-compliant parser would return identical tokens for both of those selectors, and that's indeed what `scalpel` does.

Using numbers as the attribute value is a violation of the spec, e.g., `[foo=42]`, since valid CSS identifiers cannot start with numbers. Implementing this behavior would require using a selector parser that does not conform to the selector spec.

ljharb replied:
Right - the problem is that we can't use spec-compliant selectors, because we're not selecting HTML - we're selecting from an RST, which has non-string values (HTML only has strings).

ljharb replied:
In other words, we can only use the actual selector spec if we're restricting selectors to DOM elements - so enzyme's selection is always going to have to be different than the web spec (we test react native too, for example)

aweary replied:
In that case, I can just fork `scalpel` and redefine the grammar to require unquoted values be numerical. I'm not _super_ familiar with writing parser grammars but I think I can do it 😄

ljharb replied:
it'd be even better if we could make a new package that avoids forking and instead directly depends on scalpel, but i'm not sure how hard that'd be :-/

aweary replied:
It wouldn't really be possible, this requires making a change directly to [the grammar for attribute values](https://github.com/gajus/scalpel/blob/master/src/grammar.ne#L86-L89) and there's no way to do that without changing `grammar.ne`.

aweary commented:
@ljharb here's a PR to my fork of `scalpel` that updates the grammar to require quoted string values and also parses unquoted numerical values: https://github.com/aweary/scalpel/pull/1

I was going to publish this as `rst-selector-parser` and use that in this PR instead.

aweary commented:
@ljharb I've integrated the new `rst-selector-parser` package which has actually simplified a lot of things. I updated the grammar to parse all the unquoted values we support (https://github.com/aweary/rst-selector-parser/pull/2, https://github.com/aweary/rst-selector-parser/pull/3, https://github.com/aweary/rst-selector-parser/pull/4) so the parser will return the value we want instead of a stringified version (for example, `[foo=NaN]` will return `NaN` for the token value).

aweary commented on packages/enzyme/src/Utils.js:
```diff
@@ -206,144 +205,19 @@ export function withSetStateAllowed(fn) {
-  throw new TypeError(
-    `Enzyme::Unable to parse selector '[${propName}=${propValue}]'. ` +
-    `Perhaps you forgot to escape a string? Try '[${propName}="${trimmedValue}"]' instead.`,
-  );
```
Any selector that is unparsable will fail during the parsing step now.

aweary commented:
@ljharb @lelandrichardson I think this is ready for a full review. I marked all the issues this should resolve in the top comment and added tests for them.

aweary commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+      return nodeHasProperty(node, token.name, token.value);
+    case PSEUDO_ELEMENT:
+    case PSEUDO_CLASS:
+      throw new Error('Enzyme::Selector does not support psuedo-element or psuedo-class selectors.');
```
Adding support for pseudo-selectors will be really easy. For example, implementing `:not` (https://github.com/airbnb/enzyme/pull/460). I figure this can be done in a follow-up.

ljharb replied:
Followups for these things are great; this PR's big enough already :-)

aweary commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+      return nodeHasProperty(node, token.name, token.value);
+    case PSEUDO_ELEMENT:
+    case PSEUDO_CLASS:
+      throw new Error('Enzyme::Selector does not support psuedo-element or psuedo-class selectors.');
```
For reference, I implemented `:not` with this same approach here: https://github.com/aweary/react-tester/blob/master/src/selector-parser.js#L91-L102

aweary commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+     * @example '[data-foo=foo]' matches <div data-foo="foo" />
+     */
+    case ATTRIBUTE_VALUE:
+      return nodeHasProperty(node, token.name, token.value);
```
I just realized the parser returns the operator used in attribute value selectors, so we can now support selectors like `[foo|="bar"]` and `[foo^="bar"]`

aweary commented:
@ljharb do you think we'll be able to get this in for v3?

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -13,39 +10,13 @@ import {
-    });
-
-    it('throws for malformed selectors', () => {
-      expect(() => fn('div[data-name="xyz"')).to.throw(/Enzyme::Selector received what appears to be a malformed string selector/);
```
does this no longer throw (somewhere)? it's a malformed selector.

aweary replied:
It does throw, but since the new `rst-selector-parser` manages all parsing this should be covered by [the unit tests for the package](https://github.com/aweary/rst-selector-parser/tree/master/test). I figure at this point we'd be testing the behavior of a dependency with its own tests.

But I'll add back a test that ensures that we catch the general parsing error and re-throw something more specific. Then I think it's safe to assume invalid selectors will throw, and if the `rst-selector-parser` coverage is missing a case we can add it there.

On that note, I'd be happy to include `rst-selector-parser` as part of this monorepo if we want to.

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -97,71 +68,59 @@ describe('RSTTraversal', () => {
-    it('should parse false as a literal', () => {
-      const node = $(<div foo />);
-
-      expect(nodeHasProperty(node, 'foo', 'true')).to.equal(true);
```
should this test not still be valid for `true` instead of `'true'`?

aweary replied:
It should, I'll add that case.

aweary replied:
Added test case for booleans: https://github.com/airbnb/enzyme/pull/1086/files#diff-67b0df7735c607bdef49d8a20292e57aR71

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -97,71 +68,59 @@ describe('RSTTraversal', () => {
-      expect(nodeHasProperty(<div foo={2} />, 'foo', '-0')).to.equal(false);
+      expect(nodeHasProperty(<div foo={0} />, 'foo', 0)).to.equal(true);
+      expect(nodeHasProperty(<div foo={0} />, 'foo', +0)).to.equal(true);
+      expect(nodeHasProperty(<div foo={-0} />, 'foo', -0)).to.equal(true);
```
can we add test cases that `-0` does not match `0`, and `0` does not match `-0`?

aweary replied:
Done.

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -97,71 +68,59 @@ describe('RSTTraversal', () => {
     });
 
     it('should work with empty strings', () => {
       expect(nodeHasProperty(<div foo={''} />, 'foo', '')).to.equal(true);
```
for funsies, let's add `foo=""` also?

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -97,71 +68,59 @@ describe('RSTTraversal', () => {
-
-      expect(() => nodeHasProperty(node, 'title', 'foo')).to.throw();
+      expect(nodeHasProperty(<div foo={Infinity} />, 'foo', Infinity)).to.equal(true);
+      expect(nodeHasProperty(<div foo={Infinity} />, 'foo', +Infinity)).to.equal(true);
```
let's add a case where `Infinity` fails to match a prop value of `Infinity`?

aweary replied:
Done.

ljharb replied:
It'd also be good to ensure NaN and Infinity don't equate, since both are not finite

ljharb commented on packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx:
```diff
@@ -558,18 +558,17 @@ describe('shallow', () => {
-        TypeError,
-        'Enzyme::Unable to parse selector \'[type=text]\'. Perhaps you forgot to escape a string? Try \'[type="text"]\' instead.', // eslint-disable-line max-len
-      );
+      expect(() => wrapper.find('[type=text]')).to.throw();
```
can we add more info for all these `throw` assertions? it'd be nice to assert at least on the exception type; even nicer to assert on the error message

ljharb commented on packages/enzyme-test-suite/test/selector-spec.jsx:
```diff
@@ -173,6 +173,29 @@ describe('ComplexSelector', () => {
+        const wrapper = renderMethod(<div className="foo" />);
+        ['is', 'filter', 'not', 'every'].forEach((method) => {
+          expect(() => wrapper[method]('.foo + div')).to.throw(
+            'This method does not support complex CSS selectors',
```
similarly, can all these assert on both type and message?

ljharb commented on packages/enzyme-test-suite/test/selector-spec.jsx:
```diff
@@ -188,6 +211,113 @@ describe('ComplexSelector', () => {
+        expect(wrapper.find('.foo + .foo').length).to.equal(2);
+        expect(wrapper.find('.foo + .foo').at(0).text()).to.equal('foo2');
+        expect(wrapper.find('.foo + .foo').at(1).text()).to.equal('foo3');
+        expect(wrapper.find('.foo + .foo + .foo').length).to.equal(1);
```
`.to.have.lengthOf(1)`?

ljharb commented on packages/enzyme-test-suite/test/selector-spec.jsx:
```diff
@@ -188,6 +211,113 @@ describe('ComplexSelector', () => {
+            <div className="foo">foo3</div>
+          </div>,
+        );
+        expect(wrapper.find('.foo + .foo').length).to.equal(2);
```
`.to.have.lengthOf(2)`?

ljharb commented on packages/enzyme-test-suite/test/selector-spec.jsx:
```diff
@@ -188,6 +211,113 @@ describe('ComplexSelector', () => {
+            <div data-foo-2="2" />
+          </div>,
+        );
+        expect(wrapper.find('[data-foo-1=1]').length).to.equal(2);
```
`.to.have.lengthOf` throughout?

ljharb commented on packages/enzyme/package.json:
```diff
@@ -39,6 +39,7 @@
     "object.entries": "^1.0.4",
     "raf": "^3.3.2",
+    "rst-selector-parser": "^2.2.0",
     "uuid": "^3.1.0"
```
it looks like we don't need this dependency anymore. is that accurate?

aweary replied:
Yes! I'll remove it.

ljharb commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+ * @param {Token} token 
+ */
+function nodeMatchesToken(node, token) {
+  if (node === null || typeof node === 'string') {
```
can you elaborate on why a string node wouldn't match? is there no selector that can match text?

aweary replied:
Correct, with the current grammar there's no way to match a string literal. `foo` would match an element of type `foo`. We *could* augment the grammar to allow something like `"foo"` but I'm not sure how useful that really is.

ljharb commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+ * @param {Function|Object|String} selector 
+ */
+export function buildPredicate(selector) {
+  switch (typeof selector) {
```
fwiw, this bypasses the engine optimization when `typeof foo` is repeated. This would be better as an `if`/`else`, for a number of reasons.

aweary replied:
@ljharb can you elaborate on what optimization this bypasses? I'm not familiar.

ljharb replied:
engines are able to do `typeof foo === 'some string'` checks very very fast; when you store `typeof foo` in a variable (or a binding, as here) it has to go the slower path.

ljharb commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+      // constructor
+      return node => node && node.type === selector;
+    case 'object':
+      if (!Array.isArray(selector) && selector !== null && !isEmpty(selector)) {
```
what does lodash return for `isEmpty(null)`?

if we're keeping `!== null` for perf, then we should be checking that first, before `isArray`.

aweary replied:
I'll figure out what it returns and update this. I just copied what was already there FWIW: https://github.com/airbnb/enzyme/blob/c15dd479684a9e3423d3efb5aea592fcfde1a842/packages/enzyme/src/RSTTraversal.js#L119

ljharb commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,318 @@
+export function reduceTreeBySelector(selector, wrapper) {
+  const root = wrapper.getNodeInternal();
+  let results = [];
+  switch (typeof selector) {
```
same here wrt typeof and if/else

aweary commented:
@ljharb I think I've addressed all of your feedback!

lelandrichardson commented on packages/enzyme-test-suite/test/selector-spec.jsx:
```diff
@@ -0,0 +1,324 @@
+      it('simple adjacent', () => {
+        const wrapper = renderMethod(
+          <div>
+            <div className="to-find" />
```
something worth testing is to have children of a node that are in separate arrays but adjacent to one another:

```
<div>
  <div className="a" />
  {[<div key="0" className="b" />]}
</div>
```
With this you can assert:
```
expect(wrapper.find('.a + .b'))..to.have.lengthOf(1)
```

lelandrichardson commented on packages/enzyme/src/RSTTraversal.js:
```diff
@@ -44,6 +35,31 @@ export function treeFilter(tree, fn) {
+ * @param {*} root 
+ * @param {*} targetNode 
+ */
+export function findParentNode(root, targetNode) {
```
just curious.... RSTTraversal.js has a `parentsOfNode` function that you could potentially use for this? Did it not work when you tried or did you not know about it?

aweary replied:
I did see that, and it wasn't working for me. I didn't spend a lot of time trying to fix it though since I wasn't really following how it was implemented. I'm not actually sure how it's working at all, since it calls `reverse()` on the result of `pathToNode`, but [`pathToNode` returns `null`](https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/RSTTraversal.js#L71), which is the error I was getting

```error
TypeError: Cannot read property 'reverse' of null
      at parentsOfNode (packages/enzyme/build/RSTTraversal.js:121:10)
```

It's also more complicated than we need for selector matching since AFAICT it finds all parents, not just the direct parent. If we add a `parent` pointer per discussions in https://github.com/airbnb/enzyme/issues/1085 this utility could go away altogether.

ljharb replied:
Note that that error is a pre-existing bug in enzyme 2 - #410

lelandrichardson commented on packages/enzyme/src/selectors.js:
```diff
@@ -0,0 +1,315 @@
+ * by scalpel.
+ * @param {String} selector 
+ */
+function safelyGenerateTokens(selector) {
```
feels weird to call this `safely{thing}` but still have it throw

ljharb replied:
Nah throwing is the safest thing ever! Unsafe things are what don't throw :-)

aweary replied:
Yeah, I was thinking "safe" because we explicitly throw our own error instead of letting the parser crash, and know that the tokens don't require additional validation.

lelandrichardson commented:
Oh, also, sorry this took me so long to look at!

ljharb commented on packages/enzyme-test-suite/test/RSTTraversal-spec.jsx:
```diff
@@ -97,71 +68,69 @@ describe('RSTTraversal', () => {
-      expect(nodeHasProperty(<div foo={null} />, 'foo', 'null')).to.equal(true);
-      expect(nodeHasProperty(<div foo={0} />, 'foo', 'null')).to.equal(false);
+      expect(nodeHasProperty(<div foo={null} />, 'foo', null)).to.equal(true);
+      expect(nodeHasProperty(<div foo={0} />, 'foo', null)).to.equal(false);
```
Let's also add a pair of assertions that null won't equal undefined?

aweary replied:
What's the expected behavior with `undefined`? It looks like it returns [whether `nodeProps` has an own property matching the name](https://github.com/airbnb/enzyme/blob/da9140713a946ef3bef23952e693bc917166a0b3/packages/enzyme/src/Utils.js#L355), which means `undefined` ends up matching since `foo` exists.

Is that expected?

ljharb replied:
`nodeHasProperty(<div foo={undefined} />, 'foo', undefined)` should be `true`, `nodeHasProperty(<div foo={undefined} />, 'foo', null)` should be `false`, `nodeHasProperty(<div foo={null} />, 'foo', undefined)` should be `false`.

aweary replied:
Right now `nodeHasProperty(<div foo={null} />, 'foo', undefined)` returns `true` because it's defaulting to the own-property check. I think this behavior existed before this PR too, is that a bug? I'm not sure if that will break some existing tests.

aweary replied:
ping @ljharb ^ I think that's the last point of clarification I need before this is good to go.

lelandrichardson replied:
In the interest of time, i'm going to move forward w/ this PR. I've verified that this is the current behavior. If we would like to clarify it / change it in an upcoming PR, i'm fine with that. @aweary thanks for being so thorough here.

aweary commented:
@lelandrichardson a quick heads up, to get the tests passing I had to make a change to the React 16 adapter (https://github.com/airbnb/enzyme/pull/1086/commits/a79e150df452fb30110eba49464feb6dfe905a06) since the internal instance key was changed in https://github.com/facebook/react/commit/9921e572722c0fa9d3cda65317f6de60f5d0919f

lelandrichardson commented:
@aweary sounds good. thanks.

aweary commented:
I think I've addressed all the feedback, with the exception of my question here https://github.com/airbnb/enzyme/pull/1086#discussion_r138534608 about `nodeHasProperty` behavior.

aweary commented:
It looks like the Karma tests are failing intermittently. I restarted the build a couple times on Travis and it eventually worked 🤷‍♂️

lelandrichardson commented:
@aweary do you have a link to a build where it failed? or the error message?

aweary commented:
@lelandrichardson the last build had the failure with the karma tests and React 16: https://travis-ci.org/airbnb/enzyme/jobs/274910583

It didn't mark the whole build as failed since it's considered an allowed failure, but that's the same error I was seeing before. It looks like some kind of timeout issue.
